Enable/fix QSV decoding for all QSV-supported codecs.

Originally titled, **Handbrake Nightly 2017082601 does not decode in QSV when encoder is set to H265 QSV**

### Please describe the problem or feature request in detail: 
I have a Kaby Lake & a Skylake computer, which both decode & encode H265 in QSV. However, Handbrake apparently has the decoding of HEVC shut off, even when switches are set to on when encoding HEVC with QSV. Since current & future Intel processors will be capable of this decoding, please unblock the switch to turn this decoding on.

### What are the steps to reproduce this problem: 
Turn QSV decoding on, and it doesn't work per the logs.

### What version of HandBrake you are running:
Nightly 2017082601

### What operating system and version and you running (e.g. OSX 10.11,  Windows 7, Ubuntu 14):
Windows 10

### Was there any error message or error dialog, if so please detail or provide screenshots: 
See 
"Video": {
    "Encoder": 512,
    "Level": "auto",
    "TwoPass": false,
    "Turbo": false,
    "ColorMatrixCode": 0,
    "Options": "",
    "Preset": "balanced",
    "Profile": "auto",
    "Quality": 29.0,
    "HWDecode": false,
    "QSV": {
      "Decode": true,
      "AsyncDepth": 0
    }
Then: 
decoder hevc, not QSV on another line. 
Which should look like: + decoder: hevc_qsv, as it does when going from HEVC to AVC using QSV.

### Please provide the full activity log for the encode or scan attempt. You may attach the log as a file, post a pastebin URL to the log, or place the log inline below:
```HandBrake Nightly 20170824075320-f70d349-master (2017082601) - 64bit
OS: Microsoft Windows NT 10.0.15063.0 - 64bit
CPU: Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz
Ram: 65189 MB,
GPU Information:
  Intel(R) HD Graphics 630 - 22.20.16.4749
Screen: 1600x900
Temp Dir: C:\Users\Me\AppData\Local\Temp\
Install Dir: C:\Program Files\HandBrake Nightly
Data Dir: C:\Users\Me\AppData\Roaming\HandBrake Team\HandBrake\1.1.0.0

-------------------------------------------


# Starting Encode ...

[14:25:48] hb_init: starting libhb thread
[14:25:48] 1 job(s) to process
[14:25:48] json job:
{
  "Audio": {
    "AudioList": [
      {
        "Bitrate": 64,
        "DRC": 0.0,
        "Encoder": 67108864,
        "Gain": 1.0,
        "Mixdown": 4,
        "NormalizeMixLevel": false,
        "Samplerate": 0,
        "Track": 0,
        "DitherMethod": 0
      }
    ],
    "CopyMask": [
      1073807360,
      1073743872,
      1074003968,
      1073750016,
      1090519040,
      1074790400,
      1074266112,
      1107296256
    ],
    "FallbackEncoder": 2048
  },
  "Destination": {
    "ChapterList": [
      {
        "Name": "Chapter 1"
      },
      {
        "Name": "Chapter 2"
      },
      {
        "Name": "Chapter 3"
      },
      {
        "Name": "Chapter 4"
      },
      {
        "Name": "Chapter 5"
      },
      {
        "Name": "Chapter 6"
      },
      {
        "Name": "Chapter 7"
      },
      {
        "Name": "Chapter 8"
      },
      {
        "Name": "Chapter 9"
      }
    ],
    "ChapterMarkers": true,
    "AlignAVStart": false,
    "File": "C:\\Temp1\\1080p X265 {preset}.mkv",
    "Mp4Options": {
      "IpodAtom": false,
      "Mp4Optimize": false
    },
    "Mux": 2097152
  },
  "Filters": {
    "FilterList": [
      {
        "ID": 11,
        "Settings": {
          "crop-bottom": "0",
          "crop-left": "0",
          "crop-right": "0",
          "crop-top": "0",
          "height": "1080",
          "width": "1920"
        }
      },
      {
        "ID": 6,
        "Settings": {
          "mode": "1"
        }
      }
    ]
  },
  "PAR": {
    "Num": 1,
    "Den": 1
  },
  "Metadata": {},
  "SequenceID": 0,
  "Source": {
    "Angle": 1,
    "Range": {
      "Type": "chapter",
      "Start": 1,
      "End": 9
    },
    "Title": 1,
    "Path": "C:\\Temp1\\Z\\1080p X265.mkv"
  },
  "Subtitle": {
    "Search": {
      "Burn": false,
      "Default": false,
      "Enable": false,
      "Forced": false
    },
    "SubtitleList": [
      {
        "Burn": false,
        "Default": true,
        "Forced": false,
        "ID": 1,
        "Offset": 0,
        "Track": 0
      }
    ]
  },
  "Video": {
    "Encoder": 512,
    "Level": "auto",
    "TwoPass": false,
    "Turbo": false,
    "ColorMatrixCode": 0,
    "Options": "",
    "Preset": "balanced",
    "Profile": "auto",
    "Quality": 29.0,
    "HWDecode": false,
    "QSV": {
      "Decode": true,
      "AsyncDepth": 0
    }
  }
}
[14:25:48] CPU: Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz
[14:25:48]  - Intel microarchitecture Kaby Lake
[14:25:48]  - logical processor count: 8
[14:25:48] Intel Quick Sync Video support: yes
[14:25:48]  - Intel Media SDK hardware: API 1.24 (minimum: 1.3)
[14:25:48]  - H.264 encoder: yes
[14:25:48]     - preferred implementation: hardware (any)
[14:25:48]     - capabilities (hardware):  breftype icq+la+i+downs vsinfo opt1 opt2+mbbrc+extbrc+trellis+ib_adapt+nmpslice
[14:25:48]  - H.265 encoder: yes (8bit: yes, 10bit: yes)
[14:25:48]     - preferred implementation: hardware (any)
[14:25:48]     - capabilities (hardware):  bpyramid vsinfo opt1
[14:25:48] hb_scan: path=C:\Temp1\Z\1080p X265.mkv, title_index=1
udfread ERROR: ECMA 167 Volume Recognition failed
src/libbluray/disc/disc.c:323: failed opening UDF image C:\Temp1\Z\1080p X265.mkv
src/libbluray/disc/disc.c:424: error opening file BDMV\index.bdmv
src/libbluray/disc/disc.c:424: error opening file BDMV\BACKUP\index.bdmv
[14:25:48] bd: not a bd - trying as a stream/file instead
libdvdnav: Using dvdnav version 5.0.3
libdvdread: Encrypted DVD support unavailable.
libdvdread:DVDOpenFileUDF:UDFFindFile /VIDEO_TS/VIDEO_TS.IFO failed
libdvdread:DVDOpenFileUDF:UDFFindFile /VIDEO_TS/VIDEO_TS.BUP failed
libdvdread: Can't open file VIDEO_TS.IFO.
libdvdnav: vm: failed to read VIDEO_TS.IFO
[14:25:48] dvd: not a dvd - trying as a stream/file instead
Input #0, matroska,webm, from 'C:\Temp1\Z\1080p X265.mkv':
  Duration: 01:01:37.15, start: 0.000000, bitrate: N/A
    Chapter #0:0: start 0.000000, end 5.547000
    Metadata:
      title           : Chapter 1
    Chapter #0:1: start 5.547000, end 435.852000
    Metadata:
      title           : Chapter 2
    Chapter #0:2: start 435.852000, end 536.536000
    Metadata:
      title           : Chapter 3
    Chapter #0:3: start 536.536000, end 1094.552000
    Metadata:
      title           : Chapter 4
    Chapter #0:4: start 1094.552000, end 1726.600000
    Metadata:
      title           : Chapter 5
    Chapter #0:5: start 1726.600000, end 2375.623000
    Metadata:
      title           : Chapter 6
    Chapter #0:6: start 2375.623000, end 3000.080000
    Metadata:
      title           : Chapter 7
    Chapter #0:7: start 3000.080000, end 3632.587000
    Metadata:
      title           : Chapter 8
    Chapter #0:8: start 3632.587000, end 3696.735000
    Metadata:
      title           : Chapter 9
    Stream #0:0(eng): Audio: aac (HE-AAC)
      48000 Hz, 5.1, fltp (default)
    Metadata:
      title           : Surround
    Stream #0:1(eng): Subtitle: hdmv_pgs_subtitle (default)
    Stream #0:2(eng): Video: hevc (Main 10)
      yuv420p10le, tv, 1920x1080, PAR 1:1 DAR 16:9
      23.98 fps, 1k tbn (default)
[14:25:48] scan: decoding previews for title 1
[14:25:48] scan: audio 0x0: aac, rate=48000Hz, bitrate=1 English (AAC) (5.1 ch)
[14:25:50] scan: 10 previews, 1920x1080, 23.976 fps, autocrop = 0/0/0/0, aspect 16:9, PAR 1:1
[14:25:50] libhb: scan thread found 1 valid title(s)
[14:25:50] starting job
[14:25:50] work: compression level not specified, track 1 setting compression level 10.00
[14:25:50] work: track 1, dithering not supported by codec
[14:25:50] job configuration:
[14:25:50]  * source
[14:25:50]    + C:\Temp1\Z\1080p X265.mkv
[14:25:50]    + title 1, chapter(s) 1 to 9
[14:25:50]    + container: matroska,webm
[14:25:50]  * destination
[14:25:50]    + C:\Temp1\1080p X265 {preset}.mkv
[14:25:50]    + container: Matroska (libavformat)
[14:25:50]      + chapter markers
[14:25:50]  * video track
[14:25:50]    + decoder: hevc
[14:25:50]    + filters
[14:25:50]      + Framerate Shaper (mode=1)
[14:25:50]        + frame rate: 23.976 fps -> constant 23.976 fps
[14:25:50]      + Crop and Scale (width=1920:height=1080:crop-top=0:crop-bottom=0:crop-left=0:crop-right=0)
[14:25:50]        + source: 1920 * 1080, crop (0/0/0/0): 1920 * 1080, scale: 1920 * 1080
[14:25:50]    + Output geometry
[14:25:50]      + storage dimensions: 1920 x 1080
[14:25:50]      + pixel aspect ratio: 1 : 1
[14:25:50]      + display dimensions: 1920 x 1080
[14:25:50]    + encoder: H.265 (Intel Media SDK)
[14:25:50]      + preset:  balanced
[14:25:50]      + profile: auto
[14:25:50]      + level:   auto
[14:25:50]      + quality: 29.00 (QP)
[14:25:50]  * subtitle track 1, English [PGS] (track 0, id 0x1, Picture) -> Passthrough, Default
[14:25:50]  * audio track 1
[14:25:50]    + decoder: English (AAC) (5.1 ch) (track 1, id 0x0)
[14:25:50]      + samplerate: 48000 Hz
[14:25:50]    + mixdown: Stereo
[14:25:50]    + gain: 1dB
[14:25:50]    + encoder: Opus (libopus)
[14:25:50]      + bitrate: 64 kbps, samplerate: 48000 Hz
[14:25:50]      + compression level: 10.00
[14:25:50] sync: expecting 88643 video frames
[14:25:51] encqsvInit: using encode-only path
[14:25:51] encqsvInit: H.265/HEVC Main profile @ level (null)
[14:25:51] encqsvInit: TargetUsage 4 AsyncDepth 5
[14:25:51] encqsvInit: GopRefDist 4 GopPicSize 32 NumRefFrame 3
[14:25:51] encqsvInit: BFramesMax 3 BRefType pyramid
[14:25:51] encqsvInit: RateControlMethod CQP with QPI 29 QPP 31 QPB 33
[14:25:51] sync: first pts video is 0
[14:25:51] sync: "Chapter 1" (1) at frame 1 time 0
[14:25:51] sync: first pts audio 0x0 is 0
[14:25:53] qsv_enc_init: using 'hardware (1)' implementation, API: 1.24
[14:25:55] sync: "Chapter 2" (2) at frame 134 time 499248
[14:25:57] sync: first pts subtitle 0x1 is 960660
[14:28:57] sync: "Chapter 3" (3) at frame 10451 time 39226687
[14:29:47] sync: "Chapter 4" (4) at frame 12867 time 48295747
[14:34:29] sync: "Chapter 5" (5) at frame 26246 time 98517168
[14:43:09] sync: "Chapter 6" (6) at frame 41401 time 155405250

# Encode Stopped

[14:43:36] work: average encoding speed for job is 39.917831 fps
[14:43:36] vfr: 42477 frames output, 0 dropped and 0 duped for CFR/PFR
[14:43:36] vfr: lost time: 0 (0 frames)
[14:43:36] vfr: gained time: 0 (0 frames) (0 not accounted for)
[14:43:36] aac-decoder done: 41602 frames, 0 decoder errors
[14:43:36] hevc-decoder done: 42533 frames, 0 decoder errors
[14:43:36] sync: got 42491 frames, 88643 expected
[14:43:36] sync: framerate min 23.981 fps, max 23.981 fps, avg 23.976 fps
[14:43:37] mux: track 0, 42460 frames, 254818300 bytes, 1150.27 kbps, fifo 4096
[14:43:37] mux: track 1, 88610 frames, 12721619 bytes, 57.43 kbps, fifo 8192
[14:43:37] mux: track 2, 526 frames, 6088243 bytes, 27.48 kbps, fifo 64
[14:43:37] libhb: work result = 1

# Encode Completed ...


# Encode Stopped
```

Here's the log I spoke of.

HandBrake Nightly 20170909150001-f626271-master (2017091001) - 64bit
OS: Microsoft Windows NT 10.0.15063.0 - 64bit
CPU: Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz
Ram: 65189 MB,
GPU Information:
  Intel(R) HD Graphics 630 - 22.20.16.4771
Screen: 1600x900
Temp Dir: C:\Users\Me\AppData\Local\Temp\
Install Dir: C:\Program Files\HandBrake Nightly
Data Dir: C:\Users\Me\AppData\Roaming\HandBrake Team\HandBrake\1.1.0.0

-------------------------------------------


# Starting Encode ...

[11:22:41] hb_init: starting libhb thread
[11:22:41] 1 job(s) to process
[11:22:41] json job:
{
  "Audio": {
    "AudioList": [
      {
        "Bitrate": 64,
        "DRC": 0.0,
        "Encoder": 67108864,
        "Gain": 1.0,
        "Mixdown": 5,
        "NormalizeMixLevel": false,
        "Samplerate": 0,
        "Track": 0,
        "DitherMethod": 0
      }
    ],
    "CopyMask": [
      1073807360,
      1073743872,
      1074003968,
      1073750016,
      1090519040,
      1074790400,
      1074266112,
      1107296256
    ],
    "FallbackEncoder": 2048
  },
  "Destination": {
    "ChapterList": [
      {
        "Name": "00:00:00.000"
      },
      {
        "Name": "00:03:54.359"
      },
      {
        "Name": "00:06:58.918"
      },
      {
        "Name": "00:10:08.441"
      },
      {
        "Name": "00:12:58.653"
      },
      {
        "Name": "00:15:08.658"
      },
      {
        "Name": "00:17:44.981"
      },
      {
        "Name": "00:20:42.783"
      },
      {
        "Name": "00:23:45.257"
      },
      {
        "Name": "00:25:44.668"
      },
      {
        "Name": "00:29:26.264"
      },
      {
        "Name": "00:33:55.659"
      },
      {
        "Name": "00:38:24.177"
      },
      {
        "Name": "00:42:10.194"
      },
      {
        "Name": "00:46:12.061"
      },
      {
        "Name": "00:52:17.801"
      },
      {
        "Name": "00:55:02.841"
      },
      {
        "Name": "00:57:37.621"
      },
      {
        "Name": "01:00:57.445"
      },
      {
        "Name": "01:04:30.116"
      },
      {
        "Name": "01:08:24.934"
      },
      {
        "Name": "01:11:42.089"
      },
      {
        "Name": "01:15:44.373"
      },
      {
        "Name": "01:19:03.322"
      }
    ],
    "ChapterMarkers": true,
    "AlignAVStart": false,
    "File": "C:\\Temp1\\1080p 10bit 8CH x265 HEVC {preset}.mkv",
    "Mp4Options": {
      "IpodAtom": false,
      "Mp4Optimize": false
    },
    "Mux": 2097152
  },
  "Filters": {
    "FilterList": [
      {
        "ID": 11,
        "Settings": {
          "crop-bottom": "0",
          "crop-left": "0",
          "crop-right": "0",
          "crop-top": "0",
          "height": "1040",
          "width": "1920"
        }
      },
      {
        "ID": 6,
        "Settings": {
          "mode": "1"
        }
      }
    ]
  },
  "PAR": {
    "Num": 1,
    "Den": 1
  },
  "Metadata": {},
  "SequenceID": 0,
  "Source": {
    "Angle": 1,
    "Range": {
      "Type": "chapter",
      "Start": 1,
      "End": 24
    },
    "Title": 1,
    "Path": "C:\\Temp1\\1080p 10bit 8CH x265 HEVC.mkv"
  },
  "Subtitle": {
    "Search": {
      "Burn": false,
      "Default": false,
      "Enable": false,
      "Forced": false
    },
    "SubtitleList": [
      {
        "Burn": false,
        "Default": true,
        "Forced": false,
        "ID": 1,
        "Offset": 0,
        "Track": 0
      }
    ]
  },
  "Video": {
    "Encoder": 512,
    "Level": "auto",
    "TwoPass": false,
    "Turbo": false,
    "ColorMatrixCode": 0,
    "Options": "",
    "Preset": "balanced",
    "Profile": "auto",
    "Quality": 29.0,
    "HWDecode": false,
    "QSV": {
      "Decode": true,
      "AsyncDepth": 0
    }
  }
}
[11:22:41] CPU: Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz
[11:22:41]  - Intel microarchitecture Kaby Lake
[11:22:41]  - logical processor count: 8
[11:22:41] Intel Quick Sync Video support: yes
[11:22:41]  - Intel Media SDK hardware: API 1.24 (minimum: 1.3)
[11:22:41]  - H.264 encoder: yes
[11:22:41]     - preferred implementation: hardware (any)
[11:22:41]     - capabilities (hardware):  breftype icq+la+i+downs vsinfo opt1 opt2+mbbrc+extbrc+trellis+ib_adapt+nmpslice
[11:22:41]  - H.265 encoder: yes (8bit: yes, 10bit: yes)
[11:22:41]     - preferred implementation: hardware (any)
[11:22:41]     - capabilities (hardware):  bpyramid vsinfo opt1
[11:22:41] hb_scan: path=C:\Temp1\1080p 10bit 8CH x265 HEVC.mkv, title_index=1
udfread ERROR: ECMA 167 Volume Recognition failed
src/libbluray/disc/disc.c:323: failed opening UDF image C:\Temp1\1080p 10bit 8CH x265 HEVC.mkv
src/libbluray/disc/disc.c:424: error opening file BDMV\index.bdmv
src/libbluray/disc/disc.c:424: error opening file BDMV\BACKUP\index.bdmv
[11:22:41] bd: not a bd - trying as a stream/file instead
libdvdnav: Using dvdnav version 5.0.3
libdvdread: Encrypted DVD support unavailable.
libdvdread:DVDOpenFileUDF:UDFFindFile /VIDEO_TS/VIDEO_TS.IFO failed
libdvdread:DVDOpenFileUDF:UDFFindFile /VIDEO_TS/VIDEO_TS.BUP failed
libdvdread: Can't open file VIDEO_TS.IFO.
libdvdnav: vm: failed to read VIDEO_TS.IFO
[11:22:41] dvd: not a dvd - trying as a stream/file instead
[matroska,webm @ 0000011e456b4d80] Read error at pos. 2066115623 (0x7b266c27)
Input #0, matroska,webm, from 'C:\Temp1\1080p 10bit 8CH x265 HEVC.mkv':
  Metadata:
    title           : 1080p.10bit.8CH.x265.HEVC
  Duration: 01:28:40.35, start: 0.000000, bitrate: N/A
    Chapter #0:0: start 0.000000, end 234.400000
    Metadata:
      title           : 00:00:00.000
    Chapter #0:1: start 234.400000, end 419.000000
    Metadata:
      title           : 00:03:54.359
    Chapter #0:2: start 419.000000, end 608.600000
    Metadata:
      title           : 00:06:58.918
    Chapter #0:3: start 608.600000, end 778.800000
    Metadata:
      title           : 00:10:08.441
    Chapter #0:4: start 778.800000, end 908.800000
    Metadata:
      title           : 00:12:58.653
    Chapter #0:5: start 908.800000, end 1065.000000
    Metadata:
      title           : 00:15:08.658
    Chapter #0:6: start 1065.000000, end 1242.800000
    Metadata:
      title           : 00:17:44.981
    Chapter #0:7: start 1242.800000, end 1425.400000
    Metadata:
      title           : 00:20:42.783
    Chapter #0:8: start 1425.400000, end 1544.800000
    Metadata:
      title           : 00:23:45.257
    Chapter #0:9: start 1544.800000, end 1766.400000
    Metadata:
      title           : 00:25:44.668
    Chapter #0:10: start 1766.400000, end 2035.800000
    Metadata:
      title           : 00:29:26.264
    Chapter #0:11: start 2035.800000, end 2304.200000
    Metadata:
      title           : 00:33:55.659
    Chapter #0:12: start 2304.200000, end 2530.200000
    Metadata:
      title           : 00:38:24.177
    Chapter #0:13: start 2530.200000, end 2772.200000
    Metadata:
      title           : 00:42:10.194
    Chapter #0:14: start 2772.200000, end 3138.000000
    Metadata:
      title           : 00:46:12.061
    Chapter #0:15: start 3138.000000, end 3303.000000
    Metadata:
      title           : 00:52:17.801
    Chapter #0:16: start 3303.000000, end 3457.800000
    Metadata:
      title           : 00:55:02.841
    Chapter #0:17: start 3457.800000, end 3657.600000
    Metadata:
      title           : 00:57:37.621
    Chapter #0:18: start 3657.600000, end 3870.200000
    Metadata:
      title           : 01:00:57.445
    Chapter #0:19: start 3870.200000, end 4105.000000
    Metadata:
      title           : 01:04:30.116
    Chapter #0:20: start 4105.000000, end 4302.200000
    Metadata:
      title           : 01:08:24.934
    Chapter #0:21: start 4302.200000, end 4544.400000
    Metadata:
      title           : 01:11:42.089
    Chapter #0:22: start 4544.400000, end 4743.400000
    Metadata:
      title           : 01:15:44.373
    Chapter #0:23: start 4743.400000, end 5320.400000
    Metadata:
      title           : 01:19:03.322
    Stream #0:0: Video: hevc (Main 10)
      yuv420p10le, tv, 1920x1040, PAR 1:1 DAR 24:13
      23.98 fps, 1k tbn (default)
    Stream #0:1: Audio: aac (HE-AAC)
      48000 Hz, 7.1(wide), fltp (default)
    Stream #0:2(eng): Subtitle: srt (default)
[11:22:41] scan: decoding previews for title 1
[matroska,webm @ 0000011e419bd3e0] Read error at pos. 2066115623 (0x7b266c27)
[11:22:42] scan: audio 0x1: aac, rate=48000Hz, bitrate=1 Unknown (AAC) (7.1 ch)
[matroska,webm @ 0000011e419bd3e0] Read error at pos. 2066038103 (0x7b253d57)
[11:22:42] scan: 10 previews, 1920x1040, 23.976 fps, autocrop = 0/0/0/0, aspect 16:9, PAR 1:1
[11:22:42] libhb: scan thread found 1 valid title(s)
[11:22:42] starting job
[11:22:42] work: compression level not specified, track 1 setting compression level 10.00
[11:22:42] work: track 1, dithering not supported by codec
[11:22:42] job configuration:
[11:22:42]  * source
[11:22:42]    + C:\Temp1\1080p 10bit 8CH x265 HEVC.mkv
[11:22:42]    + title 1, chapter(s) 1 to 24
[11:22:42]    + container: matroska,webm
[11:22:42]  * destination
[11:22:42]    + C:\Temp1\1080p 10bit 8CH x265 HEVC {preset}.mkv
[11:22:42]    + container: Matroska (libavformat)
[11:22:42]      + chapter markers
[11:22:42]  * video track
[11:22:42]    + decoder: hevc
[11:22:42]    + filters
[11:22:42]      + Framerate Shaper (mode=1)
[11:22:42]        + frame rate: 23.976 fps -> constant 23.976 fps
[11:22:42]      + Crop and Scale (width=1920:height=1040:crop-top=0:crop-bottom=0:crop-left=0:crop-right=0)
[11:22:42]        + source: 1920 * 1040, crop (0/0/0/0): 1920 * 1040, scale: 1920 * 1040
[11:22:42]    + Output geometry
[11:22:42]      + storage dimensions: 1920 x 1040
[11:22:42]      + pixel aspect ratio: 1 : 1
[11:22:42]      + display dimensions: 1920 x 1040
[11:22:42]    + encoder: H.265 (Intel Media SDK)
[11:22:42]      + preset:  balanced
[11:22:42]      + profile: auto
[11:22:42]      + level:   auto
[11:22:42]      + quality: 29.00 (QP)
[11:22:42]  * subtitle track 1, English [UTF-8] (track 0, id 0x2, Text) -> Passthrough, Default
[11:22:42]  * audio track 1
[11:22:42]    + decoder: Unknown (AAC) (7.1 ch) (track 1, id 0x1)
[11:22:42]      + samplerate: 48000 Hz
[11:22:42]    + mixdown: Dolby Surround
[11:22:42]    + gain: 1dB
[11:22:42]    + encoder: Opus (libopus)
[11:22:42]      + bitrate: 64 kbps, samplerate: 48000 Hz
[11:22:42]      + compression level: 10.00
[matroska,webm @ 0000011e43fc6860] Read error at pos. 2066115623 (0x7b266c27)
[11:22:42] sync: expecting 127560 video frames
[11:22:42] encqsvInit: using encode-only path
[11:22:42] encqsvInit: H.265/HEVC Main profile @ level (null)
[11:22:42] encqsvInit: TargetUsage 4 AsyncDepth 5
[11:22:42] encqsvInit: GopRefDist 4 GopPicSize 32 NumRefFrame 3
[11:22:42] encqsvInit: BFramesMax 3 BRefType pyramid
[11:22:42] encqsvInit: RateControlMethod CQP with QPI 29 QPP 31 QPB 33
[11:22:42] sync: first pts video is 0
[11:22:42] sync: "00:00:00.000" (1) at frame 1 time 0
[11:22:42] sync: first pts audio 0x1 is 2790
[11:22:42] qsv_enc_init: using 'hardware (1)' implementation, API: 1.24
[11:22:46] sync: first pts subtitle 0x2 is 1382310
[11:23:37] sync: "00:03:54.359" (2) at frame 5624 time 21107336
[11:24:21] sync: "00:06:58.918" (3) at frame 10047 time 37710172
[11:25:08] sync: "00:10:08.441" (4) at frame 14593 time 54774720
[11:25:48] sync: "00:12:58.653" (5) at frame 18676 time 70101281
[11:26:20] sync: "00:15:08.658" (6) at frame 21791 time 81794212
[11:26:58] sync: "00:17:44.981" (7) at frame 25539 time 95863267
[11:27:40] sync: "00:20:42.783" (8) at frame 29802 time 111865503
[11:28:15] reader: done. 1 scr changes
[11:28:16] work: average encoding speed for job is 98.754272 fps
[11:28:16] vfr: 32930 frames output, 0 dropped and 1 duped for CFR/PFR
[11:28:16] vfr: lost time: 0 (0 frames)
[11:28:16] vfr: gained time: 0 (0 frames) (0 not accounted for)
[11:28:16] aac-decoder done: 32195 frames, 0 decoder errors
[11:28:16] hevc-decoder done: 32929 frames, 0 decoder errors
[11:28:16] sync: got 32929 frames, 127560 expected
[11:28:16] sync: framerate min 11.905 fps, max 23.981 fps, avg 23.975 fps
[11:28:16] mux: track 0, 32930 frames, 279960737 bytes, 1630.35 kbps, fifo 4096
[11:28:16] mux: track 1, 68683 frames, 10654294 bytes, 62.05 kbps, fifo 8192
[11:28:16] mux: track 2, 464 frames, 24604 bytes, 0.14 kbps, fifo 64
[11:28:16] libhb: work result = 0

# Encode Completed ...

--- a/contrib/ffmpeg/A13-qsv-dx11.patch
+++ b/contrib/ffmpeg/A13-qsv-dx11.patch
@@ -1,17 +1,90 @@
+diff --git a/fftools/ffmpeg_opt.c b/fftools/ffmpeg_opt.c
+index f5ca18aa64..75d9b5669b 100644
+--- a/fftools/ffmpeg_opt.c
++++ b/fftools/ffmpeg_opt.c
+@@ -506,7 +506,17 @@ static int opt_init_hw_device(void *optctx, const char *opt, const char *arg)
+         printf("\n");
+         exit_program(0);
+     } else {
+-        return hw_device_init_from_string(arg, NULL);
++        HWDevice *dev;
++        int err;
++        if (!arg)
++            return AVERROR(ENOMEM);
++        err = hw_device_init_from_string(arg, &dev);
++        if (err < 0)
++            return err;
++        hw_device_ctx = av_buffer_ref(dev->device_ref);
++        if (!hw_device_ctx)
++            return AVERROR(ENOMEM);
++        return 0;
+     }
+ }
+ 
 diff --git a/libavcodec/qsv.c b/libavcodec/qsv.c
-index 986d4f6022..35879834d8 100644
+index 986d4f6022..2f01976be7 100644
 --- a/libavcodec/qsv.c
 +++ b/libavcodec/qsv.c
-@@ -328,7 +328,7 @@ load_plugin_fail:
+@@ -36,6 +36,8 @@
+ #include "avcodec.h"
+ #include "qsv_internal.h"
+ 
++#define MFX_IMPL_VIA_MASK(impl) (0x0f00 & (impl))
++
+ #if QSV_VERSION_ATLEAST(1, 12)
+ #include "mfx/mfxvp8.h"
+ #endif
+@@ -207,7 +209,7 @@ int ff_qsv_print_warning(void *log_ctx, mfxStatus err,
+     return ret;
+ }
+ 
+-static enum AVPixelFormat qsv_map_fourcc(uint32_t fourcc)
++enum AVPixelFormat ff_qsv_map_fourcc(uint32_t fourcc)
+ {
+     switch (fourcc) {
+     case MFX_FOURCC_NV12: return AV_PIX_FMT_NV12;
+@@ -245,6 +247,24 @@ int ff_qsv_find_surface_idx(QSVFramesContext *ctx, QSVFrame *frame)
+     return AVERROR_BUG;
+ }
+ 
++enum AVFieldOrder ff_qsv_map_picstruct(int mfx_pic_struct)
++{
++    enum AVFieldOrder field = AV_FIELD_UNKNOWN;
++    switch (mfx_pic_struct & 0xF) {
++    case MFX_PICSTRUCT_PROGRESSIVE:
++        field = AV_FIELD_PROGRESSIVE;
++        break;
++    case MFX_PICSTRUCT_FIELD_TFF:
++        field = AV_FIELD_TT;
++        break;
++    case MFX_PICSTRUCT_FIELD_BFF:
++        field = AV_FIELD_BB;
++        break;
++    }
++
++    return field;
++}
++
+ enum AVPictureType ff_qsv_map_pictype(int mfx_pic_type)
+ {
+     enum AVPictureType type;
+@@ -328,8 +348,13 @@ load_plugin_fail:
  int ff_qsv_init_internal_session(AVCodecContext *avctx, mfxSession *session,
                                   const char *load_plugins)
  {
 -    mfxIMPL impl   = MFX_IMPL_AUTO_ANY;
-+    mfxIMPL impl   = MFX_IMPL_AUTO_ANY | MFX_IMPL_VIA_D3D11;
-     mfxVersion ver = { { QSV_VERSION_MINOR, QSV_VERSION_MAJOR } };
+-    mfxVersion ver = { { QSV_VERSION_MINOR, QSV_VERSION_MAJOR } };
++#ifdef AVCODEC_QSV_LINUX_SESSION_HANDLE
++     mfxIMPL          impl = MFX_IMPL_AUTO_ANY;
++#else
++    mfxIMPL          impl = MFX_IMPL_AUTO_ANY | MFX_IMPL_VIA_D3D11;
++#endif
++    mfxVersion        ver = { { QSV_VERSION_MINOR, QSV_VERSION_MAJOR } };
++    mfxInitParam init_par = { MFX_IMPL_AUTO_ANY };
  
      const char *desc;
-@@ -406,6 +406,7 @@ static AVBufferRef *qsv_create_mids(AVBufferRef *hw_frames_ref)
+     int ret;
+@@ -406,6 +431,7 @@ static AVBufferRef *qsv_create_mids(AVBufferRef *hw_frames_ref)
      for (i = 0; i < nb_surfaces; i++) {
          QSVMid *mid = &mids[i];
          mid->handle        = frames_hwctx->surfaces[i].Data.MemId;
@@ -19,125 +92,542 @@ index 986d4f6022..35879834d8 100644
          mid->hw_frames_ref = hw_frames_ref1;
      }
  
-@@ -615,7 +616,14 @@ static mfxStatus qsv_frame_unlock(mfxHDL pthis, mfxMemId mid, mfxFrameData *ptr)
+@@ -500,7 +526,7 @@ static mfxStatus qsv_frame_alloc(mfxHDL pthis, mfxFrameAllocRequest *req,
+         frames_hwctx = frames_ctx->hwctx;
+ 
+         frames_ctx->format            = AV_PIX_FMT_QSV;
+-        frames_ctx->sw_format         = qsv_map_fourcc(i->FourCC);
++        frames_ctx->sw_format         = ff_qsv_map_fourcc(i->FourCC);
+         frames_ctx->width             = i->Width;
+         frames_ctx->height            = i->Height;
+         frames_ctx->initial_pool_size = req->NumFrameSuggested;
+@@ -615,7 +641,13 @@ static mfxStatus qsv_frame_unlock(mfxHDL pthis, mfxMemId mid, mfxFrameData *ptr)
  static mfxStatus qsv_frame_get_hdl(mfxHDL pthis, mfxMemId mid, mfxHDL *hdl)
  {
      QSVMid *qsv_mid = (QSVMid*)mid;
 -    *hdl = qsv_mid->handle;
-+
 +    if (qsv_mid->texture) {
-+        mfxHDLPair *pPair  =  (mfxHDLPair*)hdl;
-+        pPair->first  = qsv_mid->texture;
-+        pPair->second = qsv_mid->handle;
++        mfxHDLPair *pair  =  (mfxHDLPair*)hdl;
++        pair->first  = qsv_mid->texture;
++        pair->second = qsv_mid->handle;
 +    } else {
 +        *hdl = qsv_mid->handle;
 +    }
      return MFX_ERR_NONE;
  }
  
-@@ -624,8 +632,8 @@ int ff_qsv_init_session_device(AVCodecContext *avctx, mfxSession *psession,
- {
-     static const mfxHandleType handle_types[] = {
-         MFX_HANDLE_VA_DISPLAY,
--        MFX_HANDLE_D3D9_DEVICE_MANAGER,
-         MFX_HANDLE_D3D11_DEVICE,
-+        MFX_HANDLE_D3D9_DEVICE_MANAGER,
-     };
+@@ -630,12 +662,12 @@ int ff_qsv_init_session_device(AVCodecContext *avctx, mfxSession *psession,
      AVHWDeviceContext    *device_ctx = (AVHWDeviceContext*)device_ref->data;
      AVQSVDeviceContext *device_hwctx = device_ctx->hwctx;
+     mfxSession        parent_session = device_hwctx->session;
++    mfxHDL                    handle = NULL;
++    mfxHandleType        handle_type = MFX_IMPL_AUTO;
+ 
+     mfxSession    session;
+     mfxVersion    ver;
+     mfxIMPL       impl;
+-    mfxHDL        handle = NULL;
+-    mfxHandleType handle_type;
+     mfxStatus err;
+ 
+     int i, ret;
+@@ -647,11 +679,26 @@ int ff_qsv_init_session_device(AVCodecContext *avctx, mfxSession *psession,
+         return ff_qsv_print_error(avctx, err,
+                                   "Error querying the session attributes");
+ 
++    if (MFX_IMPL_VIA_D3D11 == MFX_IMPL_VIA_MASK(impl))
++    {
++        handle_type = MFX_HANDLE_D3D11_DEVICE;
++    }
++    else if (MFX_IMPL_VIA_D3D9 == MFX_IMPL_VIA_MASK(impl))
++    {
++        handle_type = MFX_HANDLE_D3D9_DEVICE_MANAGER;
++    }
++    else if (MFX_IMPL_VIA_VAAPI == MFX_IMPL_VIA_MASK(impl))
++    {
++        handle_type = MFX_HANDLE_VA_DISPLAY;
++    }
++
+     for (i = 0; i < FF_ARRAY_ELEMS(handle_types); i++) {
+-        err = MFXVideoCORE_GetHandle(parent_session, handle_types[i], &handle);
+-        if (err == MFX_ERR_NONE) {
+-            handle_type = handle_types[i];
+-            break;
++        if (handle_types[i] == handle_type)
++        {
++            err = MFXVideoCORE_GetHandle(parent_session, handle_types[i], &handle);
++            if (err == MFX_ERR_NONE) {
++                break;
++            }
+         }
+         handle = NULL;
+     }
 diff --git a/libavcodec/qsv_internal.h b/libavcodec/qsv_internal.h
-index b63a7d6a31..e8a766d15e 100644
+index b63a7d6a31..fe2cda4b22 100644
 --- a/libavcodec/qsv_internal.h
 +++ b/libavcodec/qsv_internal.h
-@@ -46,6 +46,8 @@ typedef struct QSVMid {
-     AVBufferRef *hw_frames_ref;
-     mfxHDL handle;
- 
-+    void *texture;
-+
+@@ -49,6 +49,11 @@ typedef struct QSVMid {
      AVFrame *locked_frame;
      AVFrame *hw_frame;
      mfxFrameSurface1 surf;
++    /**
++     * ID3D11Texture2D texture in which the frame is located for D3D11VA device. 
++     * Null in case of DXVA2 device.
++     */
++    void *texture;
+ } QSVMid;
+ 
+ typedef struct QSVFrame {
+@@ -92,9 +97,13 @@ int ff_qsv_codec_id_to_mfx(enum AVCodecID codec_id);
+ int ff_qsv_profile_to_mfx(enum AVCodecID codec_id, int profile);
+ int ff_qsv_level_to_mfx(enum AVCodecID codec_id, int level);
+ 
++enum AVPixelFormat ff_qsv_map_fourcc(uint32_t fourcc);
++
+ int ff_qsv_map_pixfmt(enum AVPixelFormat format, uint32_t *fourcc);
+ enum AVPictureType ff_qsv_map_pictype(int mfx_pic_type);
+ 
++enum AVFieldOrder ff_qsv_map_picstruct(int mfx_pic_struct);
++
+ int ff_qsv_init_internal_session(AVCodecContext *avctx, mfxSession *session,
+                                  const char *load_plugins);
+ 
+diff --git a/libavcodec/qsvdec.c b/libavcodec/qsvdec.c
+index 46aa2d6814..7e48c833ab 100644
+--- a/libavcodec/qsvdec.c
++++ b/libavcodec/qsvdec.c
+@@ -147,19 +147,21 @@ static int check_dec_param(AVCodecContext *avctx, QSVContext *q, mfxVideoParam *
+     return 1;
+ }
+ 
+-static int qsv_decode_init(AVCodecContext *avctx, QSVContext *q)
++static int qsv_decode_preinit(AVCodecContext *avctx, QSVContext *q, enum AVPixelFormat pix_fmt, mfxVideoParam *param)
+ {
+-    const AVPixFmtDescriptor *desc;
+     mfxSession session = NULL;
+     int iopattern = 0;
+-    mfxVideoParam param = { 0 };
+-    int frame_width  = avctx->coded_width;
+-    int frame_height = avctx->coded_height;
+     int ret;
++    enum AVPixelFormat pix_fmts[3] = {
++        AV_PIX_FMT_QSV, /* opaque format in case of video memory output */
++        pix_fmt,        /* system memory format obtained from bitstream parser */
++        AV_PIX_FMT_NONE };
+ 
+-    desc = av_pix_fmt_desc_get(avctx->sw_pix_fmt);
+-    if (!desc)
+-        return AVERROR_BUG;
++    ret = ff_get_format(avctx, pix_fmts);
++    if (ret < 0) {
++        q->orig_pix_fmt = avctx->pix_fmt = AV_PIX_FMT_NONE;
++        return ret;
++    }
+ 
+     if (!q->async_fifo) {
+         q->async_fifo = av_fifo_alloc(q->async_depth * qsv_fifo_item_size());
+@@ -197,54 +199,72 @@ static int qsv_decode_init(AVCodecContext *avctx, QSVContext *q)
+         return ret;
+     }
+ 
+-    ret = ff_qsv_codec_id_to_mfx(avctx->codec_id);
+-    if (ret < 0)
+-        return ret;
++    param->IOPattern   = q->iopattern;
++    param->AsyncDepth  = q->async_depth;
++    param->ExtParam    = q->ext_buffers;
++    param->NumExtParam = q->nb_ext_buffers;
+ 
+-    param.mfx.CodecId      = ret;
+-    param.mfx.CodecProfile = ff_qsv_profile_to_mfx(avctx->codec_id, avctx->profile);
+-    param.mfx.CodecLevel   = ff_qsv_level_to_mfx(avctx->codec_id, avctx->level);
+-
+-    param.mfx.FrameInfo.BitDepthLuma   = desc->comp[0].depth;
+-    param.mfx.FrameInfo.BitDepthChroma = desc->comp[0].depth;
+-    param.mfx.FrameInfo.Shift          = desc->comp[0].depth > 8;
+-    param.mfx.FrameInfo.FourCC         = q->fourcc;
+-    param.mfx.FrameInfo.Width          = frame_width;
+-    param.mfx.FrameInfo.Height         = frame_height;
+-    param.mfx.FrameInfo.ChromaFormat   = MFX_CHROMAFORMAT_YUV420;
+-
+-    switch (avctx->field_order) {
+-    case AV_FIELD_PROGRESSIVE:
+-        param.mfx.FrameInfo.PicStruct = MFX_PICSTRUCT_PROGRESSIVE;
+-        break;
+-    case AV_FIELD_TT:
+-        param.mfx.FrameInfo.PicStruct = MFX_PICSTRUCT_FIELD_TFF;
+-        break;
+-    case AV_FIELD_BB:
+-        param.mfx.FrameInfo.PicStruct = MFX_PICSTRUCT_FIELD_BFF;
+-        break;
+-    default:
+-        param.mfx.FrameInfo.PicStruct = MFX_PICSTRUCT_UNKNOWN;
+-        break;
+-    }
++    return 0;
++ }
+ 
+-    param.IOPattern   = q->iopattern;
+-    param.AsyncDepth  = q->async_depth;
+-    param.ExtParam    = q->ext_buffers;
+-    param.NumExtParam = q->nb_ext_buffers;
++static int qsv_decode_init(AVCodecContext *avctx, QSVContext *q, mfxVideoParam *param)
++{
++    int ret;
+ 
+-    if (!check_dec_param(avctx, q, &param)) {
+-        //Just give a warning instead of an error since it is still decodable possibly.
+-        av_log(avctx, AV_LOG_WARNING,
+-               "Current input bitstream is not supported by QSV decoder.\n");
+-    }
++    avctx->width        = param->mfx.FrameInfo.CropW;
++    avctx->height       = param->mfx.FrameInfo.CropH;
++    avctx->coded_width  = param->mfx.FrameInfo.Width;
++    avctx->coded_height = param->mfx.FrameInfo.Height;
++    avctx->level        = param->mfx.CodecLevel;
++    avctx->profile      = param->mfx.CodecProfile;
++    avctx->field_order  = ff_qsv_map_picstruct(param->mfx.FrameInfo.PicStruct);
++    avctx->pix_fmt      = ff_qsv_map_fourcc(param->mfx.FrameInfo.FourCC);
+ 
+-    ret = MFXVideoDECODE_Init(q->session, &param);
++    ret = MFXVideoDECODE_Init(q->session, param);
+     if (ret < 0)
+         return ff_qsv_print_error(avctx, ret,
+                                   "Error initializing the MFX video decoder");
+ 
+-    q->frame_info = param.mfx.FrameInfo;
++    q->frame_info = param->mfx.FrameInfo;
++
++    return 0;
++}
++
++static int qsv_decode_header(AVCodecContext *avctx, QSVContext *q, AVPacket *avpkt, enum AVPixelFormat pix_fmt, mfxVideoParam *param)
++{
++    int ret;
++
++    mfxBitstream bs = { 0 };
++
++    if (avpkt->size) {
++        bs.Data       = avpkt->data;
++        bs.DataLength = avpkt->size;
++        bs.MaxLength  = bs.DataLength;
++        bs.TimeStamp  = avpkt->pts;
++        if (avctx->field_order == AV_FIELD_PROGRESSIVE)
++            bs.DataFlag   |= MFX_BITSTREAM_COMPLETE_FRAME;
++    } else
++        return AVERROR_INVALIDDATA;
++
++
++    if(!q->session) {
++        ret = qsv_decode_preinit(avctx, q, pix_fmt, param);
++        if (ret < 0)
++            return ret;
++    }
++
++    ret = ff_qsv_codec_id_to_mfx(avctx->codec_id);
++    if (ret < 0)
++        return ret;
++
++    param->mfx.CodecId = ret;
++    ret = MFXVideoDECODE_DecodeHeader(q->session, &bs, param);
++    if (MFX_ERR_MORE_DATA == ret) {
++       return AVERROR(EAGAIN);
++    }
++    if (ret < 0)
++        return ff_qsv_print_error(avctx, ret,
++                "Error decoding stream header");
+ 
+     return 0;
+ }
+@@ -527,7 +547,8 @@ int ff_qsv_process_data(AVCodecContext *avctx, QSVContext *q,
+     uint8_t *dummy_data;
+     int dummy_size;
+     int ret;
+-    const AVPixFmtDescriptor *desc;
++    mfxVideoParam param = { 0 };
++    enum AVPixelFormat pix_fmt = AV_PIX_FMT_NV12;
+ 
+     if (!q->avctx_internal) {
+         q->avctx_internal = avcodec_alloc_context3(NULL);
+@@ -541,7 +562,6 @@ int ff_qsv_process_data(AVCodecContext *avctx, QSVContext *q,
+             return AVERROR(ENOMEM);
+ 
+         q->parser->flags |= PARSER_FLAG_COMPLETE_FRAMES;
+-        q->orig_pix_fmt   = AV_PIX_FMT_NONE;
+     }
+ 
+     if (!pkt->size)
+@@ -554,15 +574,23 @@ int ff_qsv_process_data(AVCodecContext *avctx, QSVContext *q,
+                      pkt->data, pkt->size, pkt->pts, pkt->dts,
+                      pkt->pos);
+ 
+-    avctx->field_order  = q->parser->field_order;
+     /* TODO: flush delayed frames on reinit */
+-    if (q->parser->format       != q->orig_pix_fmt    ||
+-        FFALIGN(q->parser->coded_width, 16)  != FFALIGN(avctx->coded_width, 16) ||
+-        FFALIGN(q->parser->coded_height, 16) != FFALIGN(avctx->coded_height, 16)) {
+-        enum AVPixelFormat pix_fmts[3] = { AV_PIX_FMT_QSV,
+-                                           AV_PIX_FMT_NONE,
+-                                           AV_PIX_FMT_NONE };
+-        enum AVPixelFormat qsv_format;
++
++    // sw_pix_fmt, coded_width/height should be set for ff_get_format(),
++    // assume sw_pix_fmt is NV12 and coded_width/height to be 1280x720,
++    // the assumption may be not corret but will be updated after header decoded if not true.
++    if (q->orig_pix_fmt != AV_PIX_FMT_NONE)
++        pix_fmt = q->orig_pix_fmt;
++    if (!avctx->coded_width)
++        avctx->coded_width = 1280;
++    if (!avctx->coded_height)
++        avctx->coded_height = 720;
++
++    ret = qsv_decode_header(avctx, q, pkt, pix_fmt, &param);
++
++    if (ret >= 0 && (q->orig_pix_fmt != ff_qsv_map_fourcc(param.mfx.FrameInfo.FourCC) ||
++        avctx->coded_width  != param.mfx.FrameInfo.Width ||
++        avctx->coded_height != param.mfx.FrameInfo.Height)) {
+         AVPacket zero_pkt = {0};
+ 
+         if (q->buffered_count) {
+@@ -571,45 +599,24 @@ int ff_qsv_process_data(AVCodecContext *avctx, QSVContext *q,
+             q->buffered_count--;
+             return qsv_decode(avctx, q, frame, got_frame, &zero_pkt);
+         }
+-
+         q->reinit_flag = 0;
+ 
+-        qsv_format = ff_qsv_map_pixfmt(q->parser->format, &q->fourcc);
+-        if (qsv_format < 0) {
+-            av_log(avctx, AV_LOG_ERROR,
+-                   "Decoding pixel format '%s' is not supported\n",
+-                   av_get_pix_fmt_name(q->parser->format));
+-            ret = AVERROR(ENOSYS);
+-            goto reinit_fail;
+-        }
++        q->orig_pix_fmt = avctx->pix_fmt = pix_fmt = ff_qsv_map_fourcc(param.mfx.FrameInfo.FourCC);
+ 
+-        q->orig_pix_fmt     = q->parser->format;
+-        avctx->pix_fmt      = pix_fmts[1] = qsv_format;
+-        avctx->width        = q->parser->width;
+-        avctx->height       = q->parser->height;
+-        avctx->coded_width  = FFALIGN(q->parser->coded_width, 16);
+-        avctx->coded_height = FFALIGN(q->parser->coded_height, 16);
+-        avctx->level        = q->avctx_internal->level;
+-        avctx->profile      = q->avctx_internal->profile;
++        avctx->coded_width  = param.mfx.FrameInfo.Width;
++        avctx->coded_height = param.mfx.FrameInfo.Height;
+ 
+-        ret = ff_get_format(avctx, pix_fmts);
++        ret = qsv_decode_preinit(avctx, q, pix_fmt, &param);
+         if (ret < 0)
+             goto reinit_fail;
++        q->initialized = 0;
++    }
+ 
+-        avctx->pix_fmt = ret;
+-
+-        desc = av_pix_fmt_desc_get(avctx->pix_fmt);
+-        if (!desc)
+-            goto reinit_fail;
+-
+-         if (desc->comp[0].depth > 8) {
+-            avctx->coded_width =  FFALIGN(q->parser->coded_width, 32);
+-            avctx->coded_height = FFALIGN(q->parser->coded_height, 32);
+-        }
+-
+-        ret = qsv_decode_init(avctx, q);
++    if (!q->initialized) {
++        ret = qsv_decode_init(avctx, q, &param);
+         if (ret < 0)
+             goto reinit_fail;
++        q->initialized = 1;
+     }
+ 
+     return qsv_decode(avctx, q, frame, got_frame, pkt);
+@@ -622,4 +629,5 @@ reinit_fail:
+ void ff_qsv_decode_flush(AVCodecContext *avctx, QSVContext *q)
+ {
+     q->orig_pix_fmt = AV_PIX_FMT_NONE;
++    q->initialized = 0;
+ }
+diff --git a/libavcodec/qsvdec.h b/libavcodec/qsvdec.h
+index 111536caba..4812fb2a6b 100644
+--- a/libavcodec/qsvdec.h
++++ b/libavcodec/qsvdec.h
+@@ -63,6 +63,8 @@ typedef struct QSVContext {
+     uint32_t fourcc;
+     mfxFrameInfo frame_info;
+ 
++    int initialized;
++
+     // options set by the caller
+     int async_depth;
+     int iopattern;
+diff --git a/libavcodec/qsvdec_h2645.c b/libavcodec/qsvdec_h2645.c
+index 9b49f5506e..eb1dc336a4 100644
+--- a/libavcodec/qsvdec_h2645.c
++++ b/libavcodec/qsvdec_h2645.c
+@@ -103,6 +103,7 @@ static av_cold int qsv_decode_init(AVCodecContext *avctx)
+         }
+     }
+ 
++    s->qsv.orig_pix_fmt = AV_PIX_FMT_NV12;
+     s->packet_fifo = av_fifo_alloc(sizeof(AVPacket));
+     if (!s->packet_fifo) {
+         ret = AVERROR(ENOMEM);
+diff --git a/libavcodec/qsvdec_other.c b/libavcodec/qsvdec_other.c
+index 03251d2c85..a6f1b88ca0 100644
+--- a/libavcodec/qsvdec_other.c
++++ b/libavcodec/qsvdec_other.c
+@@ -90,6 +90,7 @@ static av_cold int qsv_decode_init(AVCodecContext *avctx)
+     }
+ #endif
+ 
++    s->qsv.orig_pix_fmt = AV_PIX_FMT_NV12;
+     s->packet_fifo = av_fifo_alloc(sizeof(AVPacket));
+     if (!s->packet_fifo) {
+         ret = AVERROR(ENOMEM);
+diff --git a/libavfilter/qsvvpp.c b/libavfilter/qsvvpp.c
+index 8d5ff2eb65..61135541e5 100644
+--- a/libavfilter/qsvvpp.c
++++ b/libavfilter/qsvvpp.c
+@@ -32,10 +32,11 @@
+ #include "qsvvpp.h"
+ #include "video.h"
+ 
+-#define IS_VIDEO_MEMORY(mode)  (mode & (MFX_MEMTYPE_VIDEO_MEMORY_DECODER_TARGET | \
++#define IS_VIDEO_MEMORY(mode)   (mode & (MFX_MEMTYPE_VIDEO_MEMORY_DECODER_TARGET | \
+                                         MFX_MEMTYPE_VIDEO_MEMORY_PROCESSOR_TARGET))
+-#define IS_OPAQUE_MEMORY(mode) (mode & MFX_MEMTYPE_OPAQUE_FRAME)
+-#define IS_SYSTEM_MEMORY(mode) (mode & MFX_MEMTYPE_SYSTEM_MEMORY)
++#define IS_OPAQUE_MEMORY(mode)  (mode & MFX_MEMTYPE_OPAQUE_FRAME)
++#define IS_SYSTEM_MEMORY(mode)  (mode & MFX_MEMTYPE_SYSTEM_MEMORY)
++#define MFX_IMPL_VIA_MASK(impl) (0x0f00 & (impl))
+ 
+ typedef struct QSVFrame {
+     AVFrame          *frame;
+@@ -405,12 +406,12 @@ static int init_vpp_session(AVFilterContext *avctx, QSVVPPContext *s)
+     AVFilterLink                *outlink = avctx->outputs[0];
+     AVQSVFramesContext  *in_frames_hwctx = NULL;
+     AVQSVFramesContext *out_frames_hwctx = NULL;
++    mfxHandleType            handle_type = MFX_IMPL_AUTO;
+ 
+     AVBufferRef *device_ref;
+     AVHWDeviceContext *device_ctx;
+     AVQSVDeviceContext *device_hwctx;
+     mfxHDL handle;
+-    mfxHandleType handle_type;
+     mfxVersion ver;
+     mfxIMPL impl;
+     int ret, i;
+@@ -497,15 +498,30 @@ static int init_vpp_session(AVFilterContext *avctx, QSVVPPContext *s)
+         return AVERROR_UNKNOWN;
+     }
+ 
++    if (MFX_IMPL_VIA_D3D11 == MFX_IMPL_VIA_MASK(impl))
++    {
++        handle_type = MFX_HANDLE_D3D11_DEVICE;
++    }
++    else if (MFX_IMPL_VIA_D3D9 == MFX_IMPL_VIA_MASK(impl))
++    {
++        handle_type = MFX_HANDLE_D3D9_DEVICE_MANAGER;
++    }
++    else if (MFX_IMPL_VIA_VAAPI == MFX_IMPL_VIA_MASK(impl))
++    {
++        handle_type = MFX_HANDLE_VA_DISPLAY;
++    }
++
+     for (i = 0; i < FF_ARRAY_ELEMS(handle_types); i++) {
+-        ret = MFXVideoCORE_GetHandle(device_hwctx->session, handle_types[i], &handle);
+-        if (ret == MFX_ERR_NONE) {
+-            handle_type = handle_types[i];
+-            break;
++        if (handle_types[i] == handle_type)
++        {
++            ret = MFXVideoCORE_GetHandle(device_hwctx->session, handle_types[i], &handle);
++            if (ret == MFX_ERR_NONE) {
++                break;
++            }
+         }
++        handle = NULL;
+     }
+-
+-    if (ret != MFX_ERR_NONE) {
++    if (!handle) {
+         av_log(avctx, AV_LOG_ERROR, "Error getting the session handle\n");
+         return AVERROR_UNKNOWN;
+     }
 diff --git a/libavutil/hwcontext_d3d11va.c b/libavutil/hwcontext_d3d11va.c
-index 6670c47579..096770b9ce 100644
+index 6670c47579..34e1ece437 100644
 --- a/libavutil/hwcontext_d3d11va.c
 +++ b/libavutil/hwcontext_d3d11va.c
-@@ -494,12 +494,12 @@ static void d3d11va_device_uninit(AVHWDeviceContext *hwdev)
-     }
- 
-     if (device_hwctx->video_device) {
--        ID3D11VideoDevice_Release(device_hwctx->video_device);
-+        //ID3D11VideoDevice_Release(device_hwctx->video_device);
-         device_hwctx->video_device = NULL;
-     }
- 
-     if (device_hwctx->video_context) {
--        ID3D11VideoContext_Release(device_hwctx->video_context);
-+        //ID3D11VideoContext_Release(device_hwctx->video_context);
-         device_hwctx->video_context = NULL;
-     }
- 
-@@ -510,6 +510,42 @@ static void d3d11va_device_uninit(AVHWDeviceContext *hwdev)
+@@ -510,15 +510,57 @@ static void d3d11va_device_uninit(AVHWDeviceContext *hwdev)
      }
  }
  
-+static int d3d11va_device_find_qsv_adapter(AVHWDeviceContext *ctx, UINT creationFlags)
++static int d3d11va_device_find_adapter_by_vendor_id(AVHWDeviceContext *ctx, UINT creationFlags, char *vendor)
 +{
 +    HRESULT hr;
-+    IDXGIAdapter *pAdapter = NULL;
++    IDXGIAdapter *adapter = NULL;
 +    int adapter_id = 0;
-+    IDXGIFactory2 *pDXGIFactory;
-+    hr = mCreateDXGIFactory(&IID_IDXGIFactory2, (void **)&pDXGIFactory);
-+    while (IDXGIFactory2_EnumAdapters(pDXGIFactory, adapter_id++, &pAdapter) != DXGI_ERROR_NOT_FOUND)
++    IDXGIFactory2 *factory;
++    long int vendor_id = strtol(vendor, NULL, 0);
++    hr = mCreateDXGIFactory(&IID_IDXGIFactory2, (void **)&factory);
++    while (IDXGIFactory2_EnumAdapters(factory, adapter_id++, &adapter) != DXGI_ERROR_NOT_FOUND)
 +    {
-+        ID3D11Device* g_pd3dDevice = NULL;
-+        DXGI_ADAPTER_DESC adapterDesc;
++        ID3D11Device* device = NULL;
++        DXGI_ADAPTER_DESC adapter_desc;
 +
-+        hr = mD3D11CreateDevice(pAdapter, D3D_DRIVER_TYPE_UNKNOWN, NULL, creationFlags, NULL, 0, D3D11_SDK_VERSION, &g_pd3dDevice, NULL, NULL);
++        hr = mD3D11CreateDevice(adapter, D3D_DRIVER_TYPE_UNKNOWN, NULL, creationFlags, NULL, 0, D3D11_SDK_VERSION, &device, NULL, NULL);
 +        if (FAILED(hr)) {
 +            av_log(ctx, AV_LOG_ERROR, "D3D11CreateDevice returned error\n");
 +            continue;
 +        }
 +
-+        hr = IDXGIAdapter2_GetDesc(pAdapter, &adapterDesc);
++        hr = IDXGIAdapter2_GetDesc(adapter, &adapter_desc);
 +        if (FAILED(hr)) {
 +            av_log(ctx, AV_LOG_ERROR, "IDXGIAdapter2_GetDesc returned error\n");
 +            continue;
 +        }
 +
-+        if (pAdapter)
-+            IDXGIAdapter_Release(pAdapter);
++        if (device)
++            ID3D11Device_Release(device);
 +
-+        if (adapterDesc.VendorId == 0x8086) {
-+            IDXGIFactory2_Release(pDXGIFactory);
++        if (adapter)
++            IDXGIAdapter_Release(adapter);
++
++        if (adapter_desc.VendorId == vendor_id) {
++            IDXGIFactory2_Release(factory);
 +            return adapter_id - 1;
 +        }
 +    }
-+    IDXGIFactory2_Release(pDXGIFactory);
++    IDXGIFactory2_Release(factory);
 +    return -1;
 +}
 +
  static int d3d11va_device_create(AVHWDeviceContext *ctx, const char *device,
                                   AVDictionary *opts, int flags)
  {
-@@ -519,7 +555,9 @@ static int d3d11va_device_create(AVHWDeviceContext *ctx, const char *device,
+     AVD3D11VADeviceContext *device_hwctx = ctx->hwctx;
+ 
+     HRESULT hr;
++    AVDictionaryEntry *e;
      IDXGIAdapter           *pAdapter = NULL;
      ID3D10Multithread      *pMultithread;
      UINT creationFlags = D3D11_CREATE_DEVICE_VIDEO_SUPPORT;
 +    int adapter = -1;
      int is_debug       = !!av_dict_get(opts, "debug", NULL, 0);
-+    int is_qsv         = !!av_dict_get(opts, "d3d11va_qsv", NULL, 0);
      int ret;
  
-     // (On UWP we can't check this.)
-@@ -538,11 +576,22 @@ static int d3d11va_device_create(AVHWDeviceContext *ctx, const char *device,
+@@ -538,11 +580,23 @@ static int d3d11va_device_create(AVHWDeviceContext *ctx, const char *device,
          return AVERROR_UNKNOWN;
      }
  
-+    if (is_qsv) {
-+        adapter = d3d11va_device_find_qsv_adapter(ctx, creationFlags);
++    e = av_dict_get(opts, "vendor", NULL, 0);
++    if (e) {
++        adapter = d3d11va_device_find_adapter_by_vendor_id(ctx, creationFlags, e ? e->value : NULL);
 +        if (adapter < 0) {
-+            av_log(ctx, AV_LOG_ERROR, "Failed to find DX11 adapter with QSV support\n");
++            av_log(ctx, AV_LOG_ERROR, "Failed to find d3d11va adapter by vendor id\n");
 +            return AVERROR_UNKNOWN;
 +        }
 +    }
@@ -155,10 +645,14 @@ index 6670c47579..096770b9ce 100644
                  pAdapter = NULL;
              IDXGIFactory2_Release(pDXGIFactory);
 diff --git a/libavutil/hwcontext_qsv.c b/libavutil/hwcontext_qsv.c
-index 59e4ed9157..56f3ccc94a 100644
+index 59e4ed9157..270ef976d7 100644
 --- a/libavutil/hwcontext_qsv.c
 +++ b/libavutil/hwcontext_qsv.c
-@@ -30,6 +30,9 @@
+@@ -27,9 +27,13 @@
+ #include <pthread.h>
+ #endif
+ 
++#define COBJMACROS
  #if CONFIG_VAAPI
  #include "hwcontext_vaapi.h"
  #endif
@@ -168,45 +662,118 @@ index 59e4ed9157..56f3ccc94a 100644
  #if CONFIG_DXVA2
  #include "hwcontext_dxva2.h"
  #endif
-@@ -71,7 +74,7 @@ typedef struct QSVFramesContext {
-     AVBufferRef *child_frames_ref;
-     mfxFrameSurface1 *surfaces_internal;
-     int             nb_surfaces_used;
--
-+    void *texture;
-     // used in the frame allocator for non-opaque surfaces
-     mfxMemId *mem_ids;
-     // used in the opaque alloc request for opaque surfaces
-@@ -89,6 +92,9 @@ static const struct {
+@@ -44,6 +48,8 @@
+ #include "pixdesc.h"
+ #include "time.h"
+ 
++#define MFX_IMPL_VIA_MASK(impl) (0x0f00 & (impl))
++
+ typedef struct QSVDevicePriv {
+     AVBufferRef *child_device_ctx;
+ } QSVDevicePriv;
+@@ -89,6 +95,9 @@ static const struct {
  #if CONFIG_VAAPI
      { MFX_HANDLE_VA_DISPLAY,          AV_HWDEVICE_TYPE_VAAPI, AV_PIX_FMT_VAAPI },
  #endif
 +#if CONFIG_D3D11VA
-+    { MFX_HANDLE_D3D11_DEVICE, AV_HWDEVICE_TYPE_D3D11VA, AV_PIX_FMT_D3D11 },
++    { MFX_HANDLE_D3D11_DEVICE,        AV_HWDEVICE_TYPE_D3D11VA, AV_PIX_FMT_D3D11 },
 +#endif
  #if CONFIG_DXVA2
      { MFX_HANDLE_D3D9_DEVICE_MANAGER, AV_HWDEVICE_TYPE_DXVA2, AV_PIX_FMT_DXVA2_VLD },
  #endif
-@@ -229,6 +235,12 @@ static int qsv_init_child_ctx(AVHWFramesContext *ctx)
+@@ -119,25 +128,11 @@ static int qsv_device_init(AVHWDeviceContext *ctx)
+ {
+     AVQSVDeviceContext *hwctx = ctx->hwctx;
+     QSVDeviceContext       *s = ctx->internal->priv;
++    mfxHandleType handle_type = MFX_IMPL_AUTO;
+ 
+     mfxStatus err;
+     int i;
+ 
+-    for (i = 0; supported_handle_types[i].handle_type; i++) {
+-        err = MFXVideoCORE_GetHandle(hwctx->session, supported_handle_types[i].handle_type,
+-                                     &s->handle);
+-        if (err == MFX_ERR_NONE) {
+-            s->handle_type       = supported_handle_types[i].handle_type;
+-            s->child_device_type = supported_handle_types[i].device_type;
+-            s->child_pix_fmt     = supported_handle_types[i].pix_fmt;
+-            break;
+-        }
+-    }
+-    if (!s->handle) {
+-        av_log(ctx, AV_LOG_VERBOSE, "No supported hw handle could be retrieved "
+-               "from the session\n");
+-    }
+-
+     err = MFXQueryIMPL(hwctx->session, &s->impl);
+     if (err == MFX_ERR_NONE)
+         err = MFXQueryVersion(hwctx->session, &s->ver);
+@@ -146,6 +141,36 @@ static int qsv_device_init(AVHWDeviceContext *ctx)
+         return AVERROR_UNKNOWN;
+     }
+ 
++    if (MFX_IMPL_VIA_D3D11 == MFX_IMPL_VIA_MASK(s->impl))
++    {
++        handle_type = MFX_HANDLE_D3D11_DEVICE;
++    }
++    else if (MFX_IMPL_VIA_D3D9 == MFX_IMPL_VIA_MASK(s->impl))
++    {
++        handle_type = MFX_HANDLE_D3D9_DEVICE_MANAGER;
++    }
++    else if (MFX_IMPL_VIA_VAAPI == MFX_IMPL_VIA_MASK(s->impl))
++    {
++        handle_type = MFX_HANDLE_VA_DISPLAY;
++    }
++
++    for (i = 0; supported_handle_types[i].handle_type; i++) {
++        if (supported_handle_types[i].handle_type == handle_type) {
++            err = MFXVideoCORE_GetHandle(hwctx->session, supported_handle_types[i].handle_type,
++                                        &s->handle);
++            if (err == MFX_ERR_NONE) {
++                s->handle_type       = supported_handle_types[i].handle_type;
++                s->child_device_type = supported_handle_types[i].device_type;
++                s->child_pix_fmt     = supported_handle_types[i].pix_fmt;
++                break;
++            }
++        }
++    }
++    if (!s->handle) {
++        av_log(ctx, AV_LOG_VERBOSE, "No supported hw handle could be retrieved "
++               "from the session\n");
++    }
++
+     return 0;
+ }
+ 
+@@ -222,13 +247,19 @@ static int qsv_init_child_ctx(AVHWFramesContext *ctx)
+     if (!child_device_ref)
+         return AVERROR(ENOMEM);
+     child_device_ctx   = (AVHWDeviceContext*)child_device_ref->data;
+-
+ #if CONFIG_VAAPI
+     if (child_device_ctx->type == AV_HWDEVICE_TYPE_VAAPI) {
+         AVVAAPIDeviceContext *child_device_hwctx = child_device_ctx->hwctx;
          child_device_hwctx->display = (VADisplay)device_priv->handle;
      }
  #endif
 +#if CONFIG_D3D11VA
 +    if (child_device_ctx->type == AV_HWDEVICE_TYPE_D3D11VA) {
 +        AVD3D11VADeviceContext *child_device_hwctx = child_device_ctx->hwctx;
++        ID3D11Device_AddRef((ID3D11Device*)device_priv->handle);
 +        child_device_hwctx->device = (ID3D11Device*)device_priv->handle;
 +    }
 +#endif
  #if CONFIG_DXVA2
      if (child_device_ctx->type == AV_HWDEVICE_TYPE_DXVA2) {
          AVDXVA2DeviceContext *child_device_hwctx = child_device_ctx->hwctx;
-@@ -255,6 +267,15 @@ static int qsv_init_child_ctx(AVHWFramesContext *ctx)
+@@ -255,6 +286,16 @@ static int qsv_init_child_ctx(AVHWFramesContext *ctx)
      child_frames_ctx->width             = FFALIGN(ctx->width, 16);
      child_frames_ctx->height            = FFALIGN(ctx->height, 16);
  
 +#if CONFIG_D3D11VA
 +    if (child_device_ctx->type == AV_HWDEVICE_TYPE_D3D11VA) {
 +        AVD3D11VAFramesContext *child_frames_hwctx = child_frames_ctx->hwctx;
++        child_frames_hwctx->MiscFlags |= D3D11_RESOURCE_MISC_SHARED;
 +        if (hwctx->frame_type & MFX_MEMTYPE_VIDEO_MEMORY_PROCESSOR_TARGET)
 +            child_frames_hwctx->BindFlags = D3D11_BIND_RENDER_TARGET ;
 +        else
@@ -216,7 +783,7 @@ index 59e4ed9157..56f3ccc94a 100644
  #if CONFIG_DXVA2
      if (child_device_ctx->type == AV_HWDEVICE_TYPE_DXVA2) {
          AVDXVA2FramesContext *child_frames_hwctx = child_frames_ctx->hwctx;
-@@ -279,6 +300,18 @@ static int qsv_init_child_ctx(AVHWFramesContext *ctx)
+@@ -279,6 +320,18 @@ static int qsv_init_child_ctx(AVHWFramesContext *ctx)
          hwctx->frame_type = MFX_MEMTYPE_VIDEO_MEMORY_DECODER_TARGET;
      }
  #endif
@@ -225,7 +792,7 @@ index 59e4ed9157..56f3ccc94a 100644
 +        AVD3D11VAFramesContext *child_frames_hwctx = child_frames_ctx->hwctx;
 +        hwctx->texture = child_frames_hwctx->texture;
 +        for (i = 0; i < ctx->initial_pool_size; i++)
-+            s->surfaces_internal[i].Data.MemId = (mfxMemId)i;
++            s->surfaces_internal[i].Data.MemId = (mfxMemId)(int64_t)i;
 +        if (child_frames_hwctx->BindFlags == D3D11_BIND_DECODER)
 +            hwctx->frame_type = MFX_MEMTYPE_VIDEO_MEMORY_DECODER_TARGET;
 +        else
@@ -235,7 +802,98 @@ index 59e4ed9157..56f3ccc94a 100644
  #if CONFIG_DXVA2
      if (child_device_ctx->type == AV_HWDEVICE_TYPE_DXVA2) {
          AVDXVA2FramesContext *child_frames_hwctx = child_frames_ctx->hwctx;
-@@ -1074,7 +1107,7 @@ static void qsv_device_free(AVHWDeviceContext *ctx)
+@@ -421,7 +474,16 @@ static mfxStatus frame_unlock(mfxHDL pthis, mfxMemId mid, mfxFrameData *ptr)
+ 
+ static mfxStatus frame_get_hdl(mfxHDL pthis, mfxMemId mid, mfxHDL *hdl)
+ {
+-    *hdl = mid;
++    AVHWFramesContext    *ctx = pthis;
++    AVQSVFramesContext *hwctx = ctx->hwctx;
++
++    if (hwctx->texture) {
++        mfxHDLPair *pair  =  (mfxHDLPair*)hdl;
++        pair->first  = hwctx->texture;
++        pair->second = mid;
++    } else {
++        *hdl = mid;
++    }
+     return MFX_ERR_NONE;
+ }
+ 
+@@ -668,6 +730,11 @@ static int qsv_map_from(AVHWFramesContext *ctx,
+         child_data = (uint8_t*)(intptr_t)*(VASurfaceID*)surf->Data.MemId;
+         break;
+ #endif
++#if CONFIG_D3D11VA
++    case AV_HWDEVICE_TYPE_D3D11VA:
++        child_data = surf->Data.MemId;
++        break;
++#endif
+ #if CONFIG_DXVA2
+     case AV_HWDEVICE_TYPE_DXVA2:
+         child_data = surf->Data.MemId;
+@@ -972,6 +1039,27 @@ static int qsv_frames_derive_to(AVHWFramesContext *dst_ctx,
+         }
+         break;
+ #endif
++#if CONFIG_D3D11VA
++    case AV_HWDEVICE_TYPE_D3D11VA:
++        {
++            AVD3D11VAFramesContext *src_hwctx = src_ctx->hwctx;
++            s->surfaces_internal = av_mallocz_array(src_ctx->initial_pool_size,
++                                                    sizeof(*s->surfaces_internal));
++            if (!s->surfaces_internal)
++                return AVERROR(ENOMEM);
++            dst_hwctx->texture = src_hwctx->texture;
++            for (i = 0; i < src_ctx->initial_pool_size; i++) {
++                qsv_init_surface(dst_ctx, &s->surfaces_internal[i]);
++                s->surfaces_internal[i].Data.MemId = (mfxMemId)(int64_t)i;
++            }
++            dst_hwctx->nb_surfaces = src_ctx->initial_pool_size;
++            if (src_hwctx->BindFlags == D3D11_BIND_DECODER)
++                dst_hwctx->frame_type = MFX_MEMTYPE_VIDEO_MEMORY_DECODER_TARGET;
++            else
++                dst_hwctx->frame_type = MFX_MEMTYPE_VIDEO_MEMORY_PROCESSOR_TARGET;
++        }
++        break;
++#endif
+ #if CONFIG_DXVA2
+     case AV_HWDEVICE_TYPE_DXVA2:
+         {
+@@ -1009,14 +1097,26 @@ static int qsv_map_to(AVHWFramesContext *dst_ctx,
+ 
+     for (i = 0; i < hwctx->nb_surfaces; i++) {
+ #if CONFIG_VAAPI
+-        if (*(VASurfaceID*)hwctx->surfaces[i].Data.MemId ==
+-            (VASurfaceID)(uintptr_t)src->data[3])
+-            break;
++        if (AV_PIX_FMT_VAAPI == src->format) {
++            if (*(VASurfaceID*)hwctx->surfaces[i].Data.MemId ==
++                (VASurfaceID)(uintptr_t)src->data[3])
++                break;
++        }
++#endif
++#if CONFIG_D3D11VA
++        if (AV_PIX_FMT_D3D11 == src->format) {
++            if ((hwctx->texture == (ID3D11Texture2D*)(uintptr_t)src->data[0]) &&
++                ((ID3D11Texture2D*)hwctx->surfaces[i].Data.MemId ==
++                (ID3D11Texture2D*)(uintptr_t)src->data[1]))
++                break;
++        }
+ #endif
+ #if CONFIG_DXVA2
+-        if ((IDirect3DSurface9*)hwctx->surfaces[i].Data.MemId ==
+-            (IDirect3DSurface9*)(uintptr_t)src->data[3])
+-            break;
++        if (AV_PIX_FMT_DXVA2_VLD == src->format) {
++            if ((IDirect3DSurface9*)hwctx->surfaces[i].Data.MemId ==
++                (IDirect3DSurface9*)(uintptr_t)src->data[3])
++                break;
++        }
+ #endif
+     }
+     if (i >= hwctx->nb_surfaces) {
+@@ -1074,7 +1174,7 @@ static void qsv_device_free(AVHWDeviceContext *ctx)
      av_freep(&priv);
  }
  
@@ -244,7 +902,7 @@ index 59e4ed9157..56f3ccc94a 100644
  {
      static const struct {
          const char *name;
-@@ -1103,6 +1136,10 @@ static mfxIMPL choose_implementation(const char *device)
+@@ -1103,6 +1203,10 @@ static mfxIMPL choose_implementation(const char *device)
              impl = strtol(device, NULL, 0);
      }
  
@@ -255,7 +913,7 @@ index 59e4ed9157..56f3ccc94a 100644
      return impl;
  }
  
-@@ -1129,6 +1166,15 @@ static int qsv_device_derive_from_child(AVHWDeviceContext *ctx,
+@@ -1129,6 +1233,15 @@ static int qsv_device_derive_from_child(AVHWDeviceContext *ctx,
          }
          break;
  #endif
@@ -271,22 +929,97 @@ index 59e4ed9157..56f3ccc94a 100644
  #if CONFIG_DXVA2
      case AV_HWDEVICE_TYPE_DXVA2:
          {
-@@ -1231,9 +1277,12 @@ static int qsv_device_create(AVHWDeviceContext *ctx, const char *device,
-         // possible, even when multiple devices and drivers are available.
-         av_dict_set(&child_device_opts, "kernel_driver", "i915", 0);
-         av_dict_set(&child_device_opts, "driver",        "iHD",  0);
+@@ -1196,8 +1309,10 @@ fail:
+ static int qsv_device_derive(AVHWDeviceContext *ctx,
+                              AVHWDeviceContext *child_device_ctx, int flags)
+ {
+-    return qsv_device_derive_from_child(ctx, MFX_IMPL_HARDWARE_ANY,
+-                                        child_device_ctx, flags);
++    mfxIMPL impl;
++    impl = choose_implementation("hw_any", child_device_ctx->type);
++    return qsv_device_derive_from_child(ctx, impl,
++                                         child_device_ctx, flags);
+ }
+ 
+ static int qsv_device_create(AVHWDeviceContext *ctx, const char *device,
+@@ -1219,25 +1334,63 @@ static int qsv_device_create(AVHWDeviceContext *ctx, const char *device,
+     ctx->user_opaque = priv;
+     ctx->free        = qsv_device_free;
+ 
+-    e = av_dict_get(opts, "child_device", NULL, 0);
+-
+-    child_device_opts = NULL;
+-    if (CONFIG_VAAPI) {
++    e = av_dict_get(opts, "child_device_type", NULL, 0);
++    if (e) {
++        child_device_type = av_hwdevice_find_type_by_name(e ? e->value : NULL);
++        if (child_device_type == AV_HWDEVICE_TYPE_NONE) {
++            av_log(ctx, AV_LOG_ERROR, "Unknown child device type "
++                   "\"%s\".\n", e ? e->value : NULL);
++            return AVERROR(EINVAL);
++        }
++    } else if (CONFIG_VAAPI) {
+         child_device_type = AV_HWDEVICE_TYPE_VAAPI;
+-        // libmfx does not actually implement VAAPI properly, rather it
+-        // depends on the specific behaviour of a matching iHD driver when
+-        // used on recent Intel hardware.  Set options to the VAAPI device
+-        // creation so that we should pick a usable setup by default if
+-        // possible, even when multiple devices and drivers are available.
+-        av_dict_set(&child_device_opts, "kernel_driver", "i915", 0);
+-        av_dict_set(&child_device_opts, "driver",        "iHD",  0);
 -    } else if (CONFIG_DXVA2)
-+    } else if (CONFIG_D3D11VA) {
-+        child_device_type = AV_HWDEVICE_TYPE_D3D11VA;
-+        av_dict_set(&child_device_opts, "d3d11va_qsv",   "enabled",  0);
 +    } else if (CONFIG_DXVA2) {
          child_device_type = AV_HWDEVICE_TYPE_DXVA2;
 -    else {
++    } else if (CONFIG_D3D11VA) {
++        child_device_type = AV_HWDEVICE_TYPE_D3D11VA;
 +    } else {
          av_log(ctx, AV_LOG_ERROR, "No supported child device type is enabled\n");
          return AVERROR(ENOSYS);
      }
-@@ -1245,7 +1294,7 @@ static int qsv_device_create(AVHWDeviceContext *ctx, const char *device,
+ 
++    child_device_opts = NULL;
++    switch (child_device_type) {
++#if CONFIG_VAAPI
++    case AV_HWDEVICE_TYPE_VAAPI:
++        {
++            // libmfx does not actually implement VAAPI properly, rather it
++            // depends on the specific behaviour of a matching iHD driver when
++            // used on recent Intel hardware.  Set options to the VAAPI device
++            // creation so that we should pick a usable setup by default if
++            // possible, even when multiple devices and drivers are available.
++            av_dict_set(&child_device_opts, "kernel_driver", "i915", 0);
++            av_dict_set(&child_device_opts, "driver",        "iHD",  0);
++        }
++        break;
++#endif
++#if CONFIG_D3D11VA
++    case AV_HWDEVICE_TYPE_D3D11VA:
++        {
++            printf("AV_HWDEVICE_TYPE_D3D11VA has been selected\n");
++        }
++        break;
++#endif
++#if CONFIG_DXVA2
++    case AV_HWDEVICE_TYPE_DXVA2:
++        {
++            printf("AV_HWDEVICE_TYPE_DXVA2 has been selected\n");
++        }
++        break;
++#endif
++    default:
++        {
++            av_log(ctx, AV_LOG_ERROR, "No supported child device type is enabled\n");
++            return AVERROR(ENOSYS);
++        }
++        break;
++    }
++
++    e = av_dict_get(opts, "child_device", NULL, 0);
+     ret = av_hwdevice_ctx_create(&priv->child_device_ctx, child_device_type,
+                                  e ? e->value : NULL, child_device_opts, 0);
+     if (ret < 0)
+@@ -1245,7 +1398,7 @@ static int qsv_device_create(AVHWDeviceContext *ctx, const char *device,
  
      child_device = (AVHWDeviceContext*)priv->child_device_ctx->data;
  
@@ -296,14 +1029,35 @@ index 59e4ed9157..56f3ccc94a 100644
      return qsv_device_derive_from_child(ctx, impl, child_device, 0);
  }
 diff --git a/libavutil/hwcontext_qsv.h b/libavutil/hwcontext_qsv.h
-index b98d611cfc..f5a9691949 100644
+index b98d611cfc..d21d58eb61 100644
 --- a/libavutil/hwcontext_qsv.h
 +++ b/libavutil/hwcontext_qsv.h
-@@ -42,6 +42,7 @@ typedef struct AVQSVDeviceContext {
+@@ -34,6 +34,15 @@
+  */
+ typedef struct AVQSVDeviceContext {
+     mfxSession session;
++    /**
++     * Need to store actual handle type that session uses
++     * MFXVideoCORE_GetHandle() function returns mfxHandleType
++     * always equal to MFX_HANDLE_D3D9_DEVICE_MANAGER
++     * even when MFX_HANDLE_D3D11_DEVICE was set as handle before by
++     * MFXVideoCORE_SetHandle() to mfx session.
++     * Fixed already but will be available only with latest driver.
++     */
++    //mfxHandleType handle_type;
+ } AVQSVDeviceContext;
+ 
+ /**
+@@ -42,11 +51,11 @@ typedef struct AVQSVDeviceContext {
  typedef struct AVQSVFramesContext {
      mfxFrameSurface1 *surfaces;
      int            nb_surfaces;
-+    void              *texture;
- 
+-
      /**
       * A combination of MFX_MEMTYPE_* describing the frame pool.
+      */
+     int frame_type;
++    void              *texture;
+ } AVQSVFramesContext;
+ 
+ #endif /* AVUTIL_HWCONTEXT_QSV_H */

--- a/libhb/qsv_common.c
+++ b/libhb/qsv_common.c
@@ -100,6 +100,8 @@ static hb_triplet_t hb_qsv_h265_levels[] =
     { NULL,                            },
 };
 
+#define MFX_IMPL_VIA_MASK(impl) (0x0f00 & (impl))
+
 // check available Intel Media SDK version against a minimum
 #define HB_CHECK_MFX_VERSION(MFX_VERSION, MAJOR, MINOR) \
     (MFX_VERSION.Major == MAJOR  && MFX_VERSION.Minor >= MINOR)
@@ -2299,7 +2301,7 @@ extern EncQSVFramesContext hb_enc_qsv_frames_ctx;
 AVBufferRef *hb_hw_device_ctx = NULL;
 char *qsv_device = NULL;
 mfxHDL device_manager_handle = NULL;
-mfxHandleType device_handle_type = MFX_HANDLE_D3D11_DEVICE;
+mfxIMPL device_impl;
 
 #if defined(_WIN32) || defined(__MINGW32__)
 // Direct X
@@ -2667,32 +2669,36 @@ hb_buffer_t* hb_qsv_copy_frame(AVFrame *frame, hb_qsv_context *qsv_ctx)
 
     hb_qsv_get_free_surface_from_pool(0, HB_POOL_SURFACE_SIZE - HB_POOL_ENCODER_SIZE, &mid, &output_surface);
 
-    static const mfxHandleType handle_types[] = {
-        MFX_HANDLE_VA_DISPLAY,
-        MFX_HANDLE_D3D11_DEVICE,
-        MFX_HANDLE_D3D9_DEVICE_MANAGER
-    };
-
-    int i;
-
     AVHWDeviceContext    *device_ctx = (AVHWDeviceContext*)hb_hw_device_ctx->data;
     AVQSVDeviceContext *device_hwctx = device_ctx->hwctx;
     mfxSession        parent_session = device_hwctx->session;
 
     if (device_manager_handle == NULL)
     {
-        for (i = 0; i < 3; i++)
+        mfxHandleType handle_type;
+        int err = MFXQueryIMPL(parent_session, &device_impl);
+        if (err != MFX_ERR_NONE)
         {
-            int err = MFXVideoCORE_GetHandle(parent_session, handle_types[i], &device_manager_handle);
-            if (err == MFX_ERR_NONE)
-            {
-                device_handle_type = handle_types[i];
-                break;
-            }
-            device_manager_handle = NULL;
+            hb_error("hb_qsv_copy_frame: no impl could be retrieved");
+            return out;
         }
 
-        if (!device_manager_handle)
+        if (MFX_IMPL_VIA_D3D11 == MFX_IMPL_VIA_MASK(device_impl))
+        {
+            handle_type = MFX_HANDLE_D3D11_DEVICE;
+        }
+        else if (MFX_IMPL_VIA_D3D9 == MFX_IMPL_VIA_MASK(device_impl))
+        {
+            handle_type = MFX_HANDLE_D3D9_DEVICE_MANAGER;
+        }
+        else
+        {
+            hb_error("hb_qsv_copy_frame: unsupported impl");
+            return out;
+        }
+
+        err = MFXVideoCORE_GetHandle(parent_session, handle_type, &device_manager_handle);
+        if (err != MFX_ERR_NONE)
         {
             hb_error("hb_qsv_copy_frame: no supported hw handle could be retrieved "
                 "from the session\n");
@@ -2700,7 +2706,7 @@ hb_buffer_t* hb_qsv_copy_frame(AVFrame *frame, hb_qsv_context *qsv_ctx)
         }
     }
 
-    if (device_handle_type == MFX_HANDLE_D3D9_DEVICE_MANAGER)
+    if (MFX_IMPL_VIA_D3D9 == MFX_IMPL_VIA_MASK(device_impl))
     {
         IDirect3DDevice9 *pDevice = NULL;
         HANDLE handle;
@@ -2731,8 +2737,8 @@ hb_buffer_t* hb_qsv_copy_frame(AVFrame *frame, hb_qsv_context *qsv_ctx)
             hb_error("hb_qsv_copy_frame: unlock_device failded %d", result);
             return out;
         }
-    }
-    else
+    } 
+    else if (MFX_IMPL_VIA_D3D11 == MFX_IMPL_VIA_MASK(device_impl))
     {
         ID3D11DeviceContext *device_context = NULL;
         ID3D11Device *device = (ID3D11Device *)device_manager_handle;
@@ -2748,6 +2754,11 @@ hb_buffer_t* hb_qsv_copy_frame(AVFrame *frame, hb_qsv_context *qsv_ctx)
         output_surface->Data.MemId = mid;
         // copy input sufrace to sufrace from the pool
         ID3D11DeviceContext_CopySubresourceRegion(device_context, mid->texture, (uint64_t)mid->handle, 0, 0, 0, hb_enc_qsv_frames_ctx.input_texture, (uint64_t)input_surface->Data.MemId, NULL);
+    }
+    else 
+    {
+        hb_error("hb_qsv_copy_frame: incorrect mfx impl");
+        return out;
     }
 
     out->qsv_details.frame->data[3] = (uint8_t*)output_surface;
@@ -2793,6 +2804,9 @@ static int qsv_device_init(AVCodecContext *s)
         if (err < 0)
             return err;
     }
+
+    err = av_dict_set(&dict, "child_device_type", "d3d11va", 0);
+    err = av_dict_set(&dict, "vendor", "0x8086", 0);
 
     err = av_hwdevice_ctx_create(&hb_hw_device_ctx, AV_HWDEVICE_TYPE_QSV,
                                  0, dict, 0);

